### PR TITLE
[SPARK-28330][SQL][FOLLOWUP] Move the check of Offset from analysis to finsh analysis

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
https://github.com/apache/spark/pull/35975 supported ANSI SQL: result offset clause.
We make some check for offset in `CheckAnalysis`.
The behavior prevents add offset API into `Dataset`.
So this PR move the check of Offset from analysis to finsh analysis.

### Why are the changes needed?
Let we can add offset API into `Dataset`.

### Does this PR introduce _any_ user-facing change?
'No'.
New feature.

### How was this patch tested?
New tests

### CODE REVIEW

**Comment by beliefer** [comment]
ping @dtenedor cc @cloud-fan

**Review by dtenedor — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala [hunk 1]
```diff
@@ -138,3 +138,29 @@ object SpecialDatetimeValues extends Rule[LogicalPlan] {
...
+
+/**
+ * Validate whether the [[Offset]] is valid.
+ */
+object CheckOffsetOperator extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = {
+    plan.foreachUp {
+      case o if !o.isInstanceOf[GlobalLimit] && !o.isInstanceOf[LocalLimit]
+        && o.children.exists(_.isInstanceOf[Offset]) =>
+        throw new IllegalStateException(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala:150]: This is a user-facing error, we should use `AnalysisException`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala [hunk 2]
```diff
@@ -138,3 +138,29 @@ object SpecialDatetimeValues extends Rule[LogicalPlan] {
     }
   }
 }
+
+/**
+ * Validate whether the [[Offset]] is valid.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala:143]: Let's explain the reason. `Dataset` API eagerly analyzes the query, so a query plan may contain invalid `Offset` operators but it's not the final query plan that gets evaluated.

**Comment by beliefer** [comment]
Because https://github.com/apache/spark/pull/36417 merged.

**Pull request closed without merging**